src/imitation/rewards/reward_nets.py

```diff
@@ -409,6 +407,218 @@ def forward(self, state, action, next_state, done):
+        if isinstance(space, spaces.Box):
+            my_dim = space.shape[-1] if self.hwc_format else space.shape[0]
+        elif isinstance(space, spaces.Discrete):
+            my_dim = space.n
+        elif isinstance(space, spaces.MultiDiscrete):
+            my_dim = sum(space.nvec)
+        elif isinstance(space, spaces.MultiBinary):
+            my_dim = space.n if isinstance(space.n, int) else sum(space.n)
+        else:
+            raise TypeError(
+                "get_num_channels can't recognize the input space. "
+                + "Please provide one of Discrete, MultiDiscrete, MultiBinary, Box.",
+            )
+
+        return my_dim
+
+    def transpose(
+        self,
+        inp: th.Tensor,
+    ) -> th.Tensor:
+        """Transposes the state to put the channel dim before height and width."""
+        if len(inp.shape) == 3:
+            return th.permute(inp, (2, 0, 1))
+        elif len(inp.shape) == 4:
+            return th.permute(inp, (0, 3, 1, 2))
+        else:
+            raise ValueError(
+                "CnnRewardNet.transpose was given an input such that len(input.shape) "
+                + "is not 3 or 4.",
+            )
+
+    def forward(
+        self,
+        state: th.Tensor,
+        action: th.Tensor,
+        next_state: th.Tensor,
+        done: th.Tensor,
+    ) -> th.Tensor:
+        """Computes rewardNet value on input state, action, next_state, and done flag.
+
+        Takes inputs that will be used, transposes image states to (c,h,w) format if
+        needed, reshapes inputs to have compatible dimensions, concatenates them, and
+        inputs them into the CNN.
+
+        Args:
+            state: current state.
+            action: current action.
+            next_state: next state.
+            done: flag for whether the episode is over.
+
+        Returns:
+            th.Tensor: reward of the transition.
+        """
+        inputs = []
+        transpose_states = self.obs_is_image and self.hwc_format
+        transpose_acts = self.act_is_box_with_channels and self.hwc_format
+        if self.use_state:
+            state_ = self.transpose(state) if transpose_states else state
+            inputs.append(state_)
+        if self.use_action:
+            action_ = self.transpose(action) if transpose_acts else action
+            inputs.append(action_)
+        if self.use_next_state:
+            next_state_ = self.transpose(next_state) if transpose_states else next_state
+            inputs.append(next_state_)
+        if self.use_done:
+            inputs.append(done)
+
+        unsqueezed_inputs = [
+            tens if len(tens.shape) > 2 else tens.view(*tens.shape, 1, 1)
+            for tens in inputs
+        ]
+        max_height = max(unsqueezed_inputs, key=(lambda tens: tens.size(-2))).size(-2)
+        max_width = max(unsqueezed_inputs, key=(lambda tens: tens.size(-1))).size(-1)
+        boosted_inputs = []
+        for tens in unsqueezed_inputs:
```

Comment:
> To check I understand: boosting upsamples things to be the same dimension as the other CNN input, so you can feed them all through the same network?

Yes.

> Is this a standard thing that's been evaluated elsewhere?

Not to my knowledge, it just seemed like the obvious thing to do?

Reply:
My current sense is that the thing I'm doing is not that complex (especially when I get rid of the custom code for boosting 3D boxes), and less complex than the plan of feeding the states into a CNN, getting an output, concatenating the action, and putting that into an MLP, so I'm inclined to leave it be.

Reply:
> Not to my knowledge, it just seemed like the obvious thing to do?

It is a fairly natural thing to do (make everything the same shape, upsample rather than downsample so you don't throw away info). But it's one of the more complex parts of this implementation, and I suspect will have some performance overhead, so I do want to make sure it's earning its keep.

Perhaps a better thing to discuss in person?

I think the canonical think for a small discrete action space would be to just have a network that takes in image-based state and outputs a reward for every possible action: this is how deep Q-learning usually works.

I guess we could look at approaches for Q-learning with continuous action spaces and image based inputs (bit exotic, but surely someone has done this?) to get a sense of what architectures do/don't work.

Reply:
> (especially when I get rid of the custom code for boosting 3D boxes)

Happy to revisit after that's been changed, that may well simplify it.

Reply:
Code for boosting 3d boxes is now gone.

Reply:
Neither atari or procgen have continuous action spaces, but [duckietown](https://github.com/duckietown/gym-duckietown#actions) does, [CARLA](https://github.com/cjy1992/gym-carla/blob/c01ae77d2729c618504595474b744a1641ec471f/gym_carla/envs/carla_env.py#L65) can be set up with continuous or discrete actions, and [MineRL](https://minerl.readthedocs.io/en/latest/environments/index.html#action-space) has a continuous camera movement action (but all other actions are discrete).

Reply:
[This](https://arxiv.org/pdf/1509.02971.pdf#section.7) was the only paper that I could find that trained a Q-network on continuous action spaces and image inputs (but using DDPG). It also had a small CNN take in the image input and produce a vector that then got concatenated with the continuous actions before being fed into a small MLP.